ext/ffi/fast_call.rs

```diff
@@ -0,0 +1,195 @@
+            Some(counter)
+          } else {
+            None
+          }
+        }
+        _ => {
+          counter.integer += 1;
+          Some(counter)
+        }
+      })
+      .collect();
+
+    let mut ops = dynasmrt::x64::Assembler::new().unwrap();
+    if args_to_shift.is_empty() {
+      dynasm!(ops
+        ; .arch x64
+        ; xor rdi, rdi
+      );
+    } else {
+      for arg in args_to_shift {
+        match arg {
+          ArgCounter { integer: -5, .. } => dynasm!(ops
+            ; .arch x64
+            ; mov rdi, rsi
+          ),
+          ArgCounter { integer: -4, .. } => dynasm!(ops
+            ; .arch x64
+            ; mov rsi, rdx
+          ),
+          ArgCounter { integer: -3, .. } => dynasm!(ops
+            ; .arch x64
+            ; mov rdx, rcx
+          ),
+          ArgCounter { integer: -2, .. } => dynasm!(ops
+            ; .arch x64
+            ; mov rcx, r8
+          ),
+          ArgCounter { integer: -1, .. } => dynasm!(ops
+            ; .arch x64
+            ; mov r8, r9
+          ),
+          ArgCounter { integer: 0, sse } => dynasm!(ops
+            ; .arch x64
+            ; mov r9, [rsp + (1 + sse.max(0)) * 8]
+          ),
+          ArgCounter { integer, sse } => dynasm!(ops
+            ; .arch x64
+            ; mov rax, [rsp + (1 + integer + sse.max(0)) * 8]
+            ; mov [rsp + (integer + sse.max(0)) * 8], rax
+          ),
+        }
+      }
+    }
+
+    // tail-call
+    // SAFETY: stack pointer is not modified, therefore stack remains (un)aligned.
+    // Stack contains one extra parameter (the last one) when using >6 parameters, which the function will ignore
+    dynasm!(ops
+      ; .arch x64
+      ; mov rax, QWORD sym.ptr.as_ptr() as _
+      ; jmp rax
+    );
+
+    let executable_buf = ops.finalize().unwrap();
+    Self(executable_buf)
+  }
+}
+
+pub(crate) struct Template {
+  args: Box<[fast_api::Type]>,
+  ret: fast_api::CType,
+  symbol_ptr: *const c_void,
+}
+
+impl fast_api::FastFunction for Template {
+  fn function(&self) -> *const c_void {
+    self.symbol_ptr
+  }
+
+  fn args(&self) -> &'static [fast_api::Type] {
+    Box::leak(self.args.clone())
+  }
+
+  fn return_type(&self) -> fast_api::CType {
+    self.ret
+  }
+}
+
+impl From<&NativeType> for fast_api::Type {
+  fn from(native_type: &NativeType) -> Self {
+    match native_type {
+      NativeType::U8 | NativeType::U16 | NativeType::U32 => {
+        fast_api::Type::Uint32
+      }
+      NativeType::I8 | NativeType::I16 | NativeType::I32 => {
+        fast_api::Type::Int32
+      }
+      NativeType::F32 => fast_api::Type::Float32,
+      NativeType::F64 => fast_api::Type::Float64,
+      NativeType::Void => fast_api::Type::Void,
+      NativeType::I64 | NativeType::ISize => fast_api::Type::Int64,
+      NativeType::U64 | NativeType::USize => fast_api::Type::Uint64,
+      NativeType::Function | NativeType::Pointer => {
```

Comment:
Pointers are now supported as parameters as `Uint64`.

Reply:
updated branch with latest changes in main